### DISCUSSION TITLE
Some revamp for AdminerEvo

### DISCUSSION THREAD

**Discussion by LionelLaffineur** [adminerevo/adminerevo#191] [Announcements]
@gerryd, @Roy-Orbison, @peterpp, @FrancoisCapon

Dear community,

After diggings in the code, in the PRs, reading discussions and because of personal convictions, I have decided that AdminerEvo should evolve in some ways.  I'm sharing with you in the internal discussions first and will read your comments.  After some time, I will publish my plan on the public repo and ask for help to anyone interested.

1/ Too many plugins.  I think that there are too many things that have to be done via plugins.  I will take for example the need of using a plugin to connect to SQLite.  It is too heavy, AdminerEvo should be able to connect to any database without there is a need to have developer competences to write some PHP to setup plugins.

2/ Customizations.  I would like to make it easier to make AdminerEvo as you like it (as user) without you need developer skills, simply should you edit a configuration file which you can add next to adminer-xx.php or not if you want to keep it default.  This will allow to enable/disable features: some plugins will be converted to core (like connecting passwordless to SQLite) and some others will become features to enable/disable by config like TablesFilter.

3/ More user-friendliness.  I think it's possible to make AdminerEvo more user-friendly, easier to use while it doesn't become a fancy consumer application.  Config file will allow changing some usability behaviors without the need for a plugin or a skin.  I will take as example a request of changing the behavior of the tables list.  I would also like to make it configurable to have fieldsets always open like in earlier versions of Adminer when there was no fieldsets.

The list of plugins to convert to core is not yet defined and open to discussion.  The list of customizations is also TBD.  Feel free to comment.

I'm thinking of doing a new major release : 4.9.1 or 5.1.1, TBD.  And get rid of the support of PHP5.6, because it's too old and because I want to be able to use the null coalesce which I find much more readable than ternary operator.

So, if you have good reasons not to make AdminerEvo easier to deploy and customize without development skills, I'm interested to read them (while I hope there is none).
If you have a view on the new version number, I'm curious to read.

I'll create an announcement on the public repo and ask for help.  And divide the whole work to be done in a set of sub-tasks in discussions.

Thanks for reading, and thanks for commenting if you have remarks / ideas.

Regards,
L.

**Comment by Roy-Orbison** [comment]
I agree with your proposals. It would have to be a major version change if you're breaking compatibility with existing plugins, which is fine. Semver is a good thing.

Since it seems like breaking changes are on the table, I think that presents a great opportunity to fix some fundamental issues.

1.  As I wrote [here](https://github.com/adminerevo/internal-tests-and-discussions/discussions/34#discussioncomment-8175455):

	> Adminer is quite "quick and dirty" in many ways and, by default, it fails to differentiate between access to the Adminer interface itself from access to database connection specifications.

	So by default, Adminer should ask for a username + password to access it, and only then allow entering/selecting database details. **This could mitigate or resolve most security weaknesses**, such as those around Elastic. A basic config file is not a ridiculous requirement. If users absolutely need the old, insecure behaviour, then the config could say so, e.g.:
	```php
	<?php
	return [
		'adminer_auth' => false,
	];
	```
	In Adminer:
	```php
	...
	$config = (array) ((@include __DIR__ . '/adminer-config.php') ?: []);
	...
	if (array_key_exists('adminer_auth', $config) && !$config['adminer_auth']) {
		# unrestricted access to Adminer
	}
	else {
		# deny access unless $config['adminer_auth'] is an array of user => password_hash() results, and a match is found
	}
	...
	```
2.	Setting a namespace, using functions and/or a class, and avoiding global variables, would make it possible to cleanly customise Adminer & integrate it into other codebases. It would only need to bootstrap itself if it was not included by another file, e.g.:
	```php
	...
	if (get_included_files()[0] === __FILE__) {
		new AdminerEvo;
	}
	```
3.	Sensible plugin loading. As you mention, it is not user-friendly. I'm not sure whether to be more cautious, and require users to explicitly permit loading them, or to autoload all `adminer-plugin-*.php`, or similar, by default. Config options would be good, e.g.:
	```php
	<?php
	return [
		...
		# string or array of strings (cast to array for iteration)
		'plugins' => [
			'/path/to/default/adminer/plugins/*.php', # everything
			__DIR__ . 'adminer-plugin-specific.php', # specific
		],
	];
	```
	If requiring explicit permission, then provide a one-liner in a sample config so a non-developer could copy it. Globs may not be ideal as square brackets and other glob characters in actual file paths could be interpreted as patterns and be unmatchable.

Just some opinions. Take them or leave them :)

**Reply by LionelLaffineur** [reply]
> Just some opinions. Take them or leave them :)
What I like a lot in open source software is sharing ideas.  I would like to have the global knowledge of everything and have all possible ideas on any topic; but it's not the case !  What better than sharing and exchanging ?  So even if some ideas do not come to reality as proposed, they will always be read and processed as "food for thought" which will be part of the whole.  So I will take them and not leave them ;-)
Thanks for your contributions already!.
Your answer brought even more questions to me, I will formulate them in sub-discussions on the public repo to allow the world to comment.

One thing where you can perhaps suggest : what is the best way to create a kind of project with tasks and sub-tasks in github ?

**Reply by Roy-Orbison** [reply]
Have you tried the Project functionality of this private repo? I haven't used it myself, but it looks like you could put tasks into issues, then manage their progress.

**Reply by gerryd** [reply]
> Have you tried the Project functionality of this private repo? I haven't used it myself, but it looks like you could put tasks into issues, then manage their progress.

I haven't used them either so I had created a project in the demo repository: https://github.com/adminerevo/demo-repository/projects

I think this is indeed kind of what Lionel was looking for.

**Comment by gerryd** [comment]
I think what both @LionelLaffineur and @Roy-Orbison wrote makes absolute sense. I especially like the idea of introducing a namespace, it would open up quite a bit more options for integrating Adminer.

One thing I would like to note (it's not a vote against anything, just a remark): up until today, Adminer has never claimed ownership of any filename. From this change onward, it would claim `adminer-config.php` or something similar.

I would perhaps suggest to make this filename itself configurable from within the `adminer_object()` function, next to having a sane default like `adminer-config.php` for those who are already using that filename for something else. And, why not, allow setting the entire config from within that same context.

**Reply by Roy-Orbison** [reply]
You could define it with a namespaced constant, without defining `adminer_object()`, if the default was only set conditionally.

**Reply by gerryd** [reply]
> You could define it with a namespaced constant, without defining `adminer_object()`, if the default was only set conditionally.

That would be fairly simple and clean, indeed. Works for me!

**Comment by peterpp** [comment]
> 1/ Too many plugins.

Agree, many of them could be included and optionally enabled/configured.

> 2/ Customizations.
> 3/ More user-friendliness. 

Totally agree. This is also my idea for years. I would go with JSON format. File adminer.json could be auto-loaded in the same way as adminer.css and allow to configure both Adminer and optional plugins.

> And get rid of the support of PHP5.6

Compatibility with PHP 5.6 is already broken by using `??` operator. I personally would not start fixing PHP 8.3 compatibility without dropping at least PHP 5 (maybe also 7) and bumping the major version.

> So by default, Adminer should ask for a username + password to access it

I disagree here, it would be nice to let Adminer be functional without any configuration. This can be an opt-in configuration, but I'm still not sure this is something Adminer should care about. Everybody who place Adminer on a public location, is responsible for configuring appropriate authentication, for example by using HTTP Basic Auth. We use it even on servers that are accessible only by private VPN. This allow as to manage users for multiple services (not only for Adminer) in one place.

> Setting a namespace

For me, Adminer's codebase as one ugly unreadable unfriendly old-school PHP jungle (maybe I've already said that before). But who has a free time to refactor it from the ground? 😉

> Sensible plugin loading.

One way is to auto-load everything in `plugins` folder automatically by default, but let people to define the list of allowed plugins if somebody wants a restriction. I'm not sure if this will lead to some security problem, maybe yes, if another plugin like `AdminerFileUpload` will be vulnerable and allow to upload custom file anywhere.

**Reply by Roy-Orbison** [reply]
> I would go with JSON format. File adminer.json

JSON is typically served as text by a web server, which means its contents have to be actively prevented from being divulged. This is insecure by default. I like JSON but not for potentially sensitive data below the web root. Look at what happened with `.env` files that Nginx and Apache do not block by default. 

The advantage of PHP for config is that it array literals are practically as tidy as JSON, and can be linted on the same machine because PHP has to be installed for Adminer. Importantly, visiting the file will normally cause it to be executed, and it will output nothing. The bonus is that one can add logic to return whatever config needed, so is flexible for power users.

> I disagree here, it would be nice to let Adminer be functional without any configuration. This can be an opt-in configuration, but I'm still not sure this is something Adminer should care about. Everybody who place Adminer on a public location, is responsible for configuring appropriate authentication

This was the position taken by the MongoDB devs and they were lambasted for it, and changed their defaults, after thousands of unprotected databases were found on the public Internet. I, too, would like simplicity, but forcing the user to acknowledge the lack of security means better reputation and is an insignificant step for the average developer.

> One way is to auto-load everything in `plugins` folder automatically by default

Insecure by default again. It should be easy to enable this, but not easy to exploit.

Again, these are opinions based on my own experience, and you're all free to ignore them.

**Reply by peterpp** [reply]
> JSON is typically served as text by a web server

Right, my bad. It can't be placed to the public folder next to adminer.php.

> This was the position taken by the MongoDB devs and they were lambasted for it
> Insecure by default again.

OK, I understand and agree. We don't live in perfect world :)

**Reply by peterpp** [reply]
And there is another thing to consider. Right now, you can create index.php to add plugins and some stuff. But if you explicitly call adminer.php (for example https://adminer.your-service.com/adminer.php) you will get pure un-configured Adminer instance and that is also wrong in my opinion.

**Reply by LionelLaffineur** [reply]
I don't have a plan to change the fact that you can name adminer however you want.  I still like the idea to call it look-how-awesome-i-am.php like it is now ;-)

**Reply by Roy-Orbison** [reply]
> And there is another thing to consider. Right now, you can create index.php to add plugins and some stuff. But if you explicitly call adminer.php (for example https://adminer.your-service.com/adminer.php) you will get pure un-configured Adminer instance and that is also wrong in my opinion.

This would no longer be an issue if a basic config were required, even the "don't authenticate users" example, because if Adminer was accessed directly, it wouldn't include the config, so it wouldn't allow access.

This highlights an issue though, do we make the config something that Adminer pulls in, or something that pulls in Adminer, more like it is now and akin to WordPress? Perhaps it should be both, so that a default filename is sought, unless the config has already been passed to Adminer when it's instantiated.

**Comment by yurigiu** [comment]
One thing that I didn't see mentioned in the discussion, which is important to me: whatever direction you are going, **please** keep it as an easily deployable single-file. :)

**Reply by LionelLaffineur** [reply]
There is no doubt it will remain like that !

**Comment by LionelLaffineur** [comment]
May I suggest the discussion around the different topics into the related issues.
In case something is missing, let me know here underneath.

**Comment by shopx** [comment]
Seems like we will see AdminerPure in the future. I need nothing from what i read here.

**Reply by LionelLaffineur** [reply]
Well, it is an option.  You could also consider sharing what you need and contribute with us ;-)

**Reply by shopx** [reply]
I love Adminer for what it is. Not what i read here. Extra logins, required config files all this stuff. You call it AdminerEvo so it is Ok to discuss all of that. I am looking more for a AdminerLTS project. But thanks for your work up to here!

**Reply by LionelLaffineur** [reply]
I think you read too quick:
- the extra login is an idea discussed but which got critics and is not on the roadmap right now
- the config file is not due to be mandatory, it is well explained that it is fully optional, the default configuration being to keep the current flavor while we have the opportunity to change things without need of plugins.
There are many "behind the curtains" features that were improved, kept compatible.  But the goal is not to keep AdminerEvo 100% the copy of Adminer, I don't think software should stay without any evolution.
However, you are pointing out only the new features, while there has been a quite big work to be compatible with PHP 8+.

**Comment by mckaygerhard** [comment]
i am not fully agree, adminer is to be already simple, and is for managers on DBMS not for novicers.. 

questions like "where is the databases" are a sing of that! we cannot take the baby hand, they must first grow instead of pretending to run!

1. -> yeah.. there are several plugins that are already too simple and could be included
2. -> yeah.. but maybe this will break several older things as others pointed, a phone is not for management so maybe the current ui is ok
3. -> ilogical! as i said.. we cannot carry babies!
